Index: work/src/edu/arizona/kfs/module/purap/document/UaPaymentRequestDocument.java
===================================================================
--- work/src/edu/arizona/kfs/module/purap/document/UaPaymentRequestDocument.java	(revision 11062)
+++ work/src/edu/arizona/kfs/module/purap/document/UaPaymentRequestDocument.java	(revision 11063)
@@ -1,7 +1,5 @@
 package edu.arizona.kfs.module.purap.document;
 
-import org.kuali.kfs.fp.document.DisbursementVoucherConstants;
-import org.kuali.kfs.module.purap.PurapConstants;
 import org.kuali.kfs.module.purap.document.PaymentRequestDocument;
 import org.kuali.kfs.sys.KFSConstants;
 import org.kuali.kfs.sys.businessobject.Bank;
@@ -13,6 +11,7 @@
 import org.kuali.rice.kns.document.authorization.DocumentAuthorizer;
 import org.kuali.rice.kns.service.DocumentHelperService;
 import org.kuali.rice.kns.service.ParameterConstants.COMPONENT;
+import org.kuali.rice.kns.service.ParameterConstants.NAMESPACE;
 import org.kuali.rice.kns.util.GlobalVariables;
 import org.kuali.rice.kns.util.KNSConstants;
 
@@ -25,6 +24,7 @@
  * 
  * @author jonathan
  */
+@NAMESPACE( namespace = "KFS-PURAP" )
 @COMPONENT( component = "PaymentRequest" )
 public class UaPaymentRequestDocument extends PaymentRequestDocument {
     private static org.apache.log4j.Logger LOG = org.apache.log4j.Logger.getLogger(UaPaymentRequestDocument.class);
